src/iterator.c

```diff
@@ -1482,6 +1482,427 @@ int git_iterator_for_workdir_ext(
+	while (!indexfilelist_iterator__at_end(self)) {
+		const char *sz = git_vector_get(&ifi->filelist, ifi->pos_filelist_current);
+		if (git_index_snapshot_find(&ifi->pos_index, &ifi->entries, ifi->entry_srch, sz, 0, 0) == 0)
+			break;
+		ifi->pos_filelist_current++;
+	}
+	return (indexfilelist_iterator__at_end(self) ? GIT_ITEROVER : 0);
+}
+
+static void indexfilelist_iterator__free(git_iterator *self)
+{
+	indexfilelist_iterator *ifi = (indexfilelist_iterator *)self;
+	git_index_snapshot_release(&ifi->entries, ifi->index);
+	ifi->index = NULL;
+	git_vector_free(&ifi->filelist);
+}
+
+int git_iterator_for_indexfilelist(
+	git_iterator **iter,
+	git_index *index,
+	const git_strarray *paths,
+	git_iterator_flag_t flags,
+	const char *start,
+	const char *end)
+{
+	int error = 0;
+	indexfilelist_iterator *ifi = NULL;
+
+	assert(paths && paths->count > 0);
+
+	/* This iterator only returns items from the filelist
+	 * that happen to appear in the index, so we don't do
+	 * any of these variations.
+	 */
+	assert((flags & GIT_ITERATOR_INCLUDE_TREES) == 0);
+	assert((flags & GIT_ITERATOR_DONT_AUTOEXPAND) == 0);
+	assert((flags & GIT_ITERATOR_PRECOMPOSE_UNICODE) == 0);
+
+	ifi = git__calloc(1, sizeof(indexfilelist_iterator));
+	GITERR_CHECK_ALLOC(ifi);
+
+#if defined(GIT_TRACE_ITERATOR)
+	git_trace(GIT_TRACE_TRACE, "git_iterator_for_indexfilelist: [iter %p] [flags 0x%08lx] '%s' '%s'",
+			  ifi, flags,
+			  ((start) ? start : "(nil)"),
+			  ((end) ? end : "(nil)"));
+#endif
+
+	ITERATOR_BASE_INIT(ifi, indexfilelist, INDEXFILELIST, git_index_owner(index));
+	if ((error = iterator__update_ignore_case((git_iterator *)ifi, flags)) < 0)
+		goto done;
+
+	ifi->entry_srch = iterator__ignore_case(ifi) ?
+		git_index_entry_isrch : git_index_entry_srch;
+
+	/* Snapshot the current index and sort the entries as requested. */
+	ifi->index = index;
+	if ((error = git_index_snapshot_new(&ifi->entries, index)) < 0)
+		goto done;
+	git_vector_set_cmp(&ifi->entries, iterator__ignore_case(ifi) ?
+		git_index_entry_icmp : git_index_entry_cmp);
+	git_vector_sort(&ifi->entries);
+
+	if ((error = initialize_filelist_vector(&ifi->filelist,
+			iterator__ignore_case(ifi), paths)) < 0)
+		goto done;
+
+	indexfilelist_iterator__reset((git_iterator *)ifi, NULL, NULL);
+
+	*iter = (git_iterator *)ifi;
+	return 0;
+
+done:
+	git_iterator_free((git_iterator *)ifi);
+	return -1;
+}
+
+
+typedef struct {
+	git_iterator base;
+	git_iterator_callbacks cb;
+	git_vector filelist;
+	size_t pos_filelist_start;
+	size_t pos_filelist_end;
+	size_t pos_filelist_current;
+
+	git_buf buf_workdir;
+
+	git_buf buf_entry_path;
+	git_index_entry entry;
+	
+} workdirfilelist_iterator;
+
+static int workdirfilelist_stat(struct stat *st, git_iterator *self, const char *path)
+{
+	workdirfilelist_iterator *wdfi = (workdirfilelist_iterator *)self;
+	git_buf buf_fullpath = GIT_BUF_INIT;
+
+	git_buf_joinpath(&buf_fullpath, wdfi->buf_workdir.ptr, path);
+	return (p_lstat(buf_fullpath.ptr, st));
```

Comment:
This leaks the path's buffer. We need to keep track of `p_lstat`'s return code and call `git_buf_free(&buf_fullpath)`